### DISCUSSION TITLE
Extensions: Context structure and getter / setters

### DISCUSSION THREAD

**Discussion by o0Ignition0o** [apollographql/router#375] [Implementation]
This discussion will hopefully allow us to iterate on the various tradeoffs to be made, while we're designing a Context and Extensions.

Let's share the requirements we have here, and have this discussion lead to an ADR.

I have written a strawdog proposal for a context_ext model that can be found here:

https://github.com/o0Ignition0o/context_ext_strawdog/blob/main/src/main.rs

This is of course a very rough draft, but I believe it outlines how i envisioned getters / setters and the ability for Rust and non rust languages to interop with it

**Comment by o0Ignition0o** [comment]
My main concern is that the context should IMO be shareable across each plugin, and by that I mean I would even go as far as requiring any type within the context extensions to implement Serialize / Deserialize.

This wouldn't have an impact for Rust extensions / plugin builders, since they just wouldn't need to call serialize / deserialize functions, but it would force everyone to make their structures "interop ready" in a sense.

The main reason why I think this would make sense is because I believe we're building a rust router for everyone, and that means rustaceans, but also companies that aren't used to writing rust, and I want to make sure we always have them in mind when building extensibility and extensions.

Rust plugins would probably be faster, and we would skip a Serialize / Deserialize step which would be better perfomance wise, but it would still allow people to write their own plugins and "get things done".

**Comment by BrynCooke** [comment]
I'd like to know the use case for context sharing across plugins. Plugins should generally not be aware of each other, so should not need to communicate. If we can think of good use cases though then that's another matter.

**Reply by o0Ignition0o** [reply]
Oh that's interesting, I assumed it was by default, given we have this "chain of plugins" hydrating a context object.

I suspect one plugin could drive an other, eg an authentication plugin would fill a "user" context, which would then be hydrated further along the way, if the user key is already present for example. I guess what I'm trying to say is "plugins probably don't need to be aware of each other, but they might perform various things depending on, uh well.. context data 😅" 

If that's not the case, then maybe the context "key" would be like the plugin instance, so we would have this boundary 🤔

**Reply by Geal** [reply]
Examples:
- an authorization plugin that could get policies from a configuration file, and from scopes extracted by a JWT authentication plugin.
- a header filter on the subgraph requests that extracts the cache control header for the frontend response cache layer that needs them to decide on the proper value for its own header

**Reply by Geal** [reply]
it's related to another point: is an extension isolated to one hook point, or can it appear at multiple steps? If it is the latter, the context object is only needed to share between extensions (because the extension could store data on its own). If the former, then we'd need to separate some steps of an extension and have the context object carry the internal state

**Reply by BrynCooke** [reply]
Can we formalize on use calling the things that could interact at multiple steps a plugin rather than an extension? It's subtle, but for me a plugin is a unit of functionality whereas extensions add capability.

The idea was that a plugin would be aware of multiple hook points in the lifecycle. So something that puts something in context at the beginning of a request and then picks it up and does something later at the subgraph request would be a single plugin.

The previous way that people were propagating headers by putting stuff in context isn't relevant anymore due to us propagating the original request. They won't need to use a plugin for this, and can use a layer in the service definition. Header filter falls under this category, so let's concentrate on the JWT use case.

If we had a JWT plugin what would it look like? What should it be putting in context and how would the user want to deal with that data?

I could see a situation where the a JWT plugin took a callback in the constructor that would be called at the subgraph call time, This would make the JWT plugin 'self contained'. Are there any standard ways for dealing with JWT info that would translate well to yaml config though?

**Reply by Geal** [reply]
Not sure there's a standard, but [envoy](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/jwt_authn_filter#config-http-filters-jwt-authn) and [caddy](https://caddyserver.com/docs/modules/http.authentication.providers.jwt) have good examples

**Reply by BrynCooke** [reply]
Those example are super interesting. Maybe we try and implement something along the lines of the envoy plugin?

**Reply by o0Ignition0o** [reply]
This looks great! I however fail to see how it relates to context building/passing within a request ? [it looks like we're iterating through a well known set of structures](https://github.com/envoyproxy/go-control-plane/blob/dcf5642c8e54496938e0311fe9c48e39b609e583/envoy/extensions/filters/http/jwt_authn/v3/config.pb.validate.go#L969)

**Reply by BrynCooke** [reply]
I guess I'm arguing that that iteration is performed by the JWT plugin itself, so no need to share. 

Similarly, thinking ahead a WASM plugin would only store Value or other similar types in context.

**Reply by BrynCooke** [reply]
It would mean that a WASM plugin would not be able to take advantage of any JWT processing done in a JWT plugin, but I see this as a good thing.
The plugin system currently proposed only allows one instance of a plugin, so it's not as if you could have multiple instances of a WASM plugin one after another doing chains of stuff.

**Reply by o0Ignition0o** [reply]
Hmm that makes sense, I can't think of an actual use case that would require this, so maybe we can start without, and see if there's a need for that eventually. I guess this would solve the `hashmap` issue, like there's no hashmap at all, just a per plugin context ?

**Reply by BrynCooke** [reply]
Yeah, this would be a nice API.

**Reply by Geal** [reply]
> Those example are super interesting. Maybe we try and implement something along the lines of the envoy plugin?

we can implement something, but we should not make it available so soon, I'd like us to explore in depth what we need on authenticationand authorization first, because we have more complex use cases than HTTP reverse proxys

**Reply by Geal** [reply]
honestly I am a bit worried that making one plugin appear at multiple hook points will be hard to integrate in the router (except maybe if the entire state of the plugin is stored in the context)

**Reply by Geal** [reply]
> It would mean that a WASM plugin would not be able to take advantage of any JWT processing done in a JWT plugin, but I see this as a good thing.

That would result in plugins covering more and more feature space, instead of doing one small thing. Example: the JWT plugin verifies the signature, checks revocation, extract claims. Does it also handle authorization at the entire request level? Does it embed authorization policies at the field level? If we have per subgraph policies we also need that JWT plugin to be called at subgraph request time.
While when we have plugins communicate with each other, we get:
- JWT plugin verifies the signature, checks revocation, extract claims, stores claims and JWT in context
- authorization plugin looks at claims to accept or not the request. When the response comes back, it filters fields according to policies depending on the claims
- subgraph auth plugin can accept or redirect a query depending on the claims, and replace the JWT with another one

If plugins can share info through the context, every one of those steps can be replaced (ex: making a different authz plugin based on OPA or Casbin) without having to rewrite the entire chain

**Reply by BrynCooke** [reply]
I see your point. Cross plugin communication seems desirable.

**Reply by o0Ignition0o** [reply]
Now that’s a super nice use case to focus on, thanks for sharing!
